src/perl/Permabit/BlockDevice/VDO.pm

```diff
@@ -589,30 +592,37 @@ sub installModule {
       $self->uninstallModule($machineName);
     }
   }
-
+  # If useUpstreamModule is set, we need to use $VDO_LINUXNEXT_MODNAME(dm-vdo)
+  # module otherwise we load the $VDO_MODNAME(kvdo)
+  if ($self->{useUpstreamModule}) {
+    $moduleName = $VDO_LINUXNEXT_MODNAME;
+  }
   # Build the kernel and user tool modules from source.
-  $log->info("Installing module $moduleName $version on $machineName");
   my $module
     = Permabit::KernelModule->new(
                                   machine         => $self->getMachine(),
                                   modDir          => $self->getModuleSourceDir(),
                                   modFileName     => "kmod-$moduleName",
                                   modName         => $moduleName,
                                   modVersion      => $version,
+                                  upstreamModName => $VDO_LINUXNEXT_MODNAME,
                                   useDistribution => $self->{useDistribution},
+                                  useUpstream     => $self->{useUpstreamModule},
                                  );
   $module->load();
   $self->{_modules}{$machineName}{$moduleName} = $module;
   $self->_setModuleParameters();
 
-  $log->info("Installing module $VDO_USER_MODNAME $version on $machineName");
   my $userModule
     = Permabit::UserModule->new(
                                 machine         => $self->getMachine(),
                                 modDir          => $self->getModuleSourceDir(),
                                 modName         => $VDO_USER_MODNAME,
-                                modVersion      => $version,
+                                modVersion      => $self->{useUpstreamModule}
+                                                     ? ""
+                                                     : $self->getModuleVersion(),
```

Comment:
Again, this logic should be put in getModuleVersion().
Hm, looking at how we use the version number in the hash, the upstream case should probably have a special tag like "upstream" to differentiate it from the module we build ourselves.

Actually, for the user module, don't you need to pass a real version number?